[SPARK-16380][SQL][Example]:Update SQL examples and programming guide for Python language binding

wangmiao1981 opened the pull request:
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)
In current sql-programming-guide.md, Python examples are hard coded in the md file.

I update the file by adding a separate SparkSQLExample.py as ml examples.

In this file, I included all working and hard-coded examples as a self-contained application, except for Hive examples. For example, spark.refershtable, which doesn't exist in SparkSession. We can revisit these examples and put it in the self-contained application.
## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
Manual tests:
./bin/spark-submit examples/src/main/python/SparkSQLExample.py
Build docs and check generated document including correct examples as ml document.

wangmiao1981 committed:
add SQL example in python

wangmiao1981 committed:
add examples and change md file

wangmiao1981 commented:
@liancheng Can you review it? Thanks!

SparkQA commented:
**[Test build #61939 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61939/consoleFull)** for PR 14098 at commit [`94df090`](https://github.com/apache/spark/commit/94df0908d17d28eb6c0223456df7daad24899e47).

SparkQA commented:
**[Test build #61939 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61939/consoleFull)** for PR 14098 at commit [`94df090`](https://github.com/apache/spark/commit/94df0908d17d28eb6c0223456df7daad24899e47).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61939/
Test FAILed.

wangmiao1981 committed:
fix python style error

SparkQA commented:
**[Test build #61940 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61940/consoleFull)** for PR 14098 at commit [`d92d933`](https://github.com/apache/spark/commit/d92d933937571b65670a2a308c936c0d9061b382).

SparkQA commented:
**[Test build #61940 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61940/consoleFull)** for PR 14098 at commit [`d92d933`](https://github.com/apache/spark/commit/d92d933937571b65670a2a308c936c0d9061b382).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61940/
Test PASSed.

liancheng commented on examples/src/main/python/SparkSQLExample.py:
```diff
@@ -0,0 +1,208 @@
+    # Justin
+
+    # Select everybody, but increment the age by 1
+    df.select(df['name'], df['age'] + 1).show()
```
@rxin @cloud-fan Shall we deprecate, or at least not recommend, using `df['...']` to reference columns since this may lead to potential ambiguity in self-join cases? I'm thinking about replacing it with `col('...')` in all Python example code.

liancheng replied:
cc @JoshRosen @davies

liancheng commented on examples/src/main/python/SparkSQLExample.py:
```diff
@@ -0,0 +1,208 @@
+        .builder\
+        .appName("PythonSQL")\
+        .config("spark.some.config.option", "some-value")\
+        .getOrCreate()
```
Nit: Please use 2-space indentation here.

liancheng commented:
Thanks for doing this! Overall it's pretty nice. A few high level comments:

1. It might be better to split the whole example file into several methods, as what #14119 did. In this way, the code can be easier to follow, and we can avoid using variable names like `df1`, `df2`, and `df3`.
2. Could you please add actual output after each `.show()` call? I think you can simply copy them from #14119.
3. Could you please update example snippet label names to match those used in #14119?

wangmiao1981 commented:
@liancheng Thanks for your review! I will address your comments asap. Currently, I am working on a ML wrapper for R.

wangmiao1981 committed:
address review comments, part 1, not completed, WIP

wangmiao1981 commented:
Not completed. Please hold on for review.

SparkQA commented:
**[Test build #62405 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62405/consoleFull)** for PR 14098 at commit [`8ed04fa`](https://github.com/apache/spark/commit/8ed04fae588be39bd4e1a7f082c13b104db1fe21).

SparkQA commented:
**[Test build #62405 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62405/consoleFull)** for PR 14098 at commit [`8ed04fa`](https://github.com/apache/spark/commit/8ed04fae588be39bd4e1a7f082c13b104db1fe21).
 * This patch **fails Python style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62405/
Test FAILed.

wangmiao1981 committed:
fix python style error

SparkQA commented:
**[Test build #62413 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62413/consoleFull)** for PR 14098 at commit [`8d94dd3`](https://github.com/apache/spark/commit/8d94dd39fa6a3da3089f4474c136269910c72ff2).

SparkQA commented:
**[Test build #62413 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62413/consoleFull)** for PR 14098 at commit [`8d94dd3`](https://github.com/apache/spark/commit/8d94dd39fa6a3da3089f4474c136269910c72ff2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62413/
Test PASSed.

wangmiao1981 committed:
update examples

SparkQA commented:
**[Test build #62415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62415/consoleFull)** for PR 14098 at commit [`91a2e10`](https://github.com/apache/spark/commit/91a2e1067b24df48f25bb540248d27bd9cc57f9c).

wangmiao1981 committed:
move SQL example into sql folder

SparkQA commented:
**[Test build #62416 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62416/consoleFull)** for PR 14098 at commit [`68e65fa`](https://github.com/apache/spark/commit/68e65fa5675eed0359e7d8c36020cabc9b2cea47).

SparkQA commented:
**[Test build #62415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62415/consoleFull)** for PR 14098 at commit [`91a2e10`](https://github.com/apache/spark/commit/91a2e1067b24df48f25bb540248d27bd9cc57f9c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62415/
Test PASSed.

SparkQA commented:
**[Test build #62416 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62416/consoleFull)** for PR 14098 at commit [`68e65fa`](https://github.com/apache/spark/commit/68e65fa5675eed0359e7d8c36020cabc9b2cea47).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62416/
Test PASSed.

liancheng commented on docs/sql-programming-guide.md:
```diff
@@ -79,7 +79,7 @@ The entry point into all functionality in Spark is the [`SparkSession`](api/java
 The entry point into all functionality in Spark is the [`SparkSession`](api/python/pyspark.sql.html#pyspark.sql.SparkSession) class. To create a basic `SparkSession`, just use `SparkSession.builder`:
 
-{% include_example init_session python/sql.py %}
+{% include_example init_session python/SparkSQLExample.py %}
```
The original file name `sql.py` should be fine. Actually, the new name `SparkSQLExample.py` doesn't conform to Python convention. (You may check other file names under `examples/src/main/python`.)

liancheng replied:
And the file path is wrong. The actual path is `python/sql/SparkSqlExample.py`.

liancheng commented:
@wangmiao1981 Is this ready for review now? Also, please update the PR title to:

```
[SPARK-16380][SQL][EXAMPLE] Update SQL examples and programming guide for Python language binding
```

liancheng commented:
@wangmiao1981 I guess it's not ready yet. You may put a `[WIP]` tag in the PR title when it's in WIP status and remove it when it is ready for review.

rxin commented:
Please be careful with case sensitivity. It broke the release candidate last time.

liancheng commented:
Yea, especially on case insensitive OS'es like Mac and Windows, the doc actually builds successfully even when cases of the example file names don't match. I guess that's probably why we missed SPARK-16553.

wangmiao1981 commented:
@liancheng Sorry for replying late. I was on vacation last a few days.

I have addressed most of your comments. Only the .md file is not updated yet. 

By the way, I am trying to make the hive example work, but I still can not get it work. Any suggestions? I found that pyspark sql is different from the corresponding scala hive example.

Thanks!

Miao

wangmiao1981 committed:
update md file

SparkQA commented:
**[Test build #62648 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62648/consoleFull)** for PR 14098 at commit [`ac47d8d`](https://github.com/apache/spark/commit/ac47d8d4ba2e80b4e5fdac9f586a7ccf583fab90).

wangmiao1981 committed:
rename file

SparkQA commented:
**[Test build #62649 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62649/consoleFull)** for PR 14098 at commit [`95b16f5`](https://github.com/apache/spark/commit/95b16f5bba48455af738a364861cba7ba56a791a).

SparkQA commented:
**[Test build #62648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62648/consoleFull)** for PR 14098 at commit [`ac47d8d`](https://github.com/apache/spark/commit/ac47d8d4ba2e80b4e5fdac9f586a7ccf583fab90).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62648/
Test PASSed.

wangmiao1981 committed:
change help message

SparkQA commented:
**[Test build #62649 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62649/consoleFull)** for PR 14098 at commit [`95b16f5`](https://github.com/apache/spark/commit/95b16f5bba48455af738a364861cba7ba56a791a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #62651 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62651/consoleFull)** for PR 14098 at commit [`8563ecb`](https://github.com/apache/spark/commit/8563ecb7d980cde7ccb515ed2e75da929d233568).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62649/
Test PASSed.

wangmiao1981 commented:
@liancheng I addressed all your comments. Except: 1). 2-spaces indents; I tried it, but it failed on python style tests. So I leave it 4-spaces indents; 2) `col('...')` I haven't changed it yet. Do you have a finial decision on this part? 

It is ready to review now.

Thanks!

SparkQA commented:
**[Test build #62651 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62651/consoleFull)** for PR 14098 at commit [`8563ecb`](https://github.com/apache/spark/commit/8563ecb7d980cde7ccb515ed2e75da929d233568).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62651/
Test PASSed.

liancheng commented:
@wangmiao1981 Thanks for working on this. For the Hive example, I guess you probably forgot to call `enableHiveSupport()` over the `SparkSession` object. And I made a mistake about the indentation, 4-space indentation is OK for Python. Sorry for the trouble...

I just opened PR #14317 based on your work. Sorry that I didn't realize that you finished working on this PR before sending #14317. However, that one also addressed more minor styling issues and added a separate Hive example file. Would you mind to help review it? I will attribute this work to you in #14317.

wangmiao1981 commented:
@liancheng Thanks! I will review the PR #14317

wangmiao1981 commented:
As #14317 has been merged, I close this PR.
